### ISSUE TITLE
Robust Scaling result is different in ML.NET compared to scikit-learn

### ISSUE THREAD

**Issue by JJJWD** [dotnet/machinelearning#6341]
### System information

- **OS version/distro**: Windows 10 
- **.NET Version (eg., dotnet --info)**:  .NET Framework 4.7.2
- ML.NET v1.7.1
- python v 3.8.10
- scikit-learn v1.1.2

### Issue

- **What did you do?**
I'm trying to convert my ML code in python using scikit-learn to C# using ML.NET
- **What happened?**
I noticed that the result with robust scaling in the pipeline using ML.NET did not match the results from python. The difference seems to be due to a different 'scale' value and the used function.
- **What did you expect?**
I have checked  the scale and offset of the data in EXCEL using the QUARTILE.INC() function and these results suggest that the values from python are correct. 
I've also derived from the [documentation ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.ml.transforms.normalizingtransformer.affinenormalizermodelparameters-1?view=ml-dotnet#properties) that ML.NET seems to use the `(input - offset) * scale` function to calculate the normalized value while the python code uses `(input - offset) \ scale`

- I would like to know why this result is so different and whether this is bug or not.

Below you can find an example code to demonstrate the difference
### Source code C#
```
private class TestInput
{
	[ColumnName(@"Test")]
	public float Test { get; set; }
}

[TestMethod]
public void NormalizedTest()
{
	List<TestInput> testInputs = new List<TestInput>()
	{
		new TestInput(){Test = 2.114265f },
		new TestInput(){Test = 2.837738f },
		new TestInput(){Test = 2.909677f },
		new TestInput(){Test = 3.769787f },
	};

	var mlContext = new MLContext();
	IDataView dataView = mlContext.Data.LoadFromEnumerable<TestInput>(testInputs);

	//normalize
	var normalize = mlContext.Transforms.NormalizeRobustScaling(new[] {
			new InputOutputColumnPair(@"Test", @"Test") });
	var normalizeTransform = normalize.Fit(dataView);
	var transformedData = normalizeTransform.Transform(dataView);
	var affineParams = normalizeTransform.GetNormalizerModelParameters(0) as AffineNormalizerModelParameters<float>;
	var offset = affineParams.Offset;
	var scale = affineParams.Scale;
	Console.WriteLine($"Offset: {offset}");
	Console.WriteLine($"Scale: {scale}");
	List<TestInput> vals = mlContext.Data.CreateEnumerable<TestInput>(
	transformedData, reuseRowObject: false).ToList();
	for (int i = 0; i < vals.Count; i++) {
		//y = (x - ({ offset})) * { scale}
		Console.WriteLine($"{(testInputs[i].Test - offset) * scale} \t  {vals[i].Test}");
	}
}
```

**Output**
```
Offset: 2.873708
Scale: 1.208078
-0.917466 	  -0.917466
-0.04345396  	  -0.04345396
0.04345396 	  0.04345396
1.082534 	  1.082534
```

### Source code python
```
import numpy as np
from sklearn.preprocessing import RobustScaler

vals= [2.114265, 2.837738,2.909677,3.769787]
X = np.array(vals)
scaler = RobustScaler()
y= scaler.fit_transform(X.reshape(-1, 1))
print(f'Offset {scaler.center_}')
print(f'Scale {scaler.scale_}')

print(y)
for val in vals:
    print((val-scaler.center_)/scaler.scale_)
```

**Output**
```
Offset [2.8737075]
Scale [0.46783475]
[[-1.62331357]
 [-0.07688505]
 [ 0.07688505]
 [ 1.9153761 ]]
[-1.62331357]
[-0.07688505]
[0.07688505]
[1.9153761]
```

**Comment by LittleLittleCloud** [comment]
It's because of different way how sklearn and ml.net calculate `scale`.

scale in mlnet is calculated using the below fomula
`scale = 1 / ((column.max - column.min) * (quantilemax - quantilemin) / 100)`

in your example, it's 
`1 / ((3.769 - 2.114) * (75 - 25) / 100) = 1.208`

in sklearn, the fomula is
`scale = percentile(column, quantilemax) - percentile(column, quantilemin)`

where percentile is [np.percentile](https://numpy.org/doc/stable/reference/generated/numpy.percentile.html#r08bde0ebf37b-1), which computes the q-th percentile of data along an array.

In your example, the 25%-th of `X` is (2.114265 + (0.333 - 0.25) * (2.837 - 2.114265) = 2.6568. The 75% of `X` is 3.1247. And the difference is 0.4678, as per your output shows.

Note that there're two differences on how scale is calculated in ml.net
- `scale` in ml.net is `1/scale` in sklearn. that explains the difference on how normalized data get calculated.
- the way how percentile is calculated in ml.net assumes that data points are evenly distributed, **which I would say is not correct**. For example, the percentage(50) for [0, 1, 2, 3] and [0, 0, 0, 3] should be different, but according to the formula in ml.net, they are the same.

To verify it, you can provide mlnet and sklearn with an evenly distributed array, where mlnet's scale should be 1 over sklearn's scale.

**Comment by JJJWD** [comment]
@LittleLittleCloud Thanks for your explanation. I’ve tried your suggestion of testing it with an evenly distributed array of [0, 1, 2, 3] and the result in ML.NET is indeed 1/sklearn’s scale. 
I do wonder whether the robust scaler in ML.NET should also provide the option to use percentile method of calculating the scale despite the costs since it does seem to slightly influence the results of prediction model. However, for now I’m satisfied to know the answer and to know that my code conversion is correct.
